tools/resourcedocsgen/cmd/docs/registry.go

```diff
@@ -39,6 +40,76 @@ import (
 	concpool "github.com/sourcegraph/conc/pool"
 )
 
+// PackageMetadataProvider is an interface for providers that retrieve package metadata
+// from either the filesystem directory or the Pulumi Registry API.
+type PackageMetadataProvider interface {
+	// GetPackageMetadata returns metadata for a specific package
+	GetPackageMetadata(pkgName string) (*pkg.PackageMeta, error)
+	// ListPackageMetadata returns metadata for all packages
+	ListPackageMetadata() ([]*pkg.PackageMeta, error)
+}
+
+// FileSystemProvider implements PackageMetadataProvider using the local yaml data files
+// in the pulumi/registry repository.
+type FileSystemProvider struct {
+	registryDir string
+}
+
+// RegistryAPIProvider implements PackageMetadataProvider using the Pulumi API
+// to retrieve package metadata.
+type RegistryAPIProvider struct {
+	apiURL string
+}
+
+// PackageMetadata represents the API response structure for package metadata
+// from the Pulumi Registry API.
+// TODO: import this from the pulumi-service if possible.
+type PackageMetadata struct {
+	Name          string    `json:"name"`
+	Publisher     string    `json:"publisher"`
+	Source        string    `json:"source"`
+	Version       string    `json:"version"`
+	Title         string    `json:"title,omitempty"`
+	Description   string    `json:"description,omitempty"`
+	LogoURL       string    `json:"logoUrl,omitempty"`
+	RepoURL       string    `json:"repoUrl,omitempty"`
+	Category      string    `json:"category,omitempty"`
+	IsFeatured    bool      `json:"isFeatured"`
+	PackageTypes  []string  `json:"packageTypes,omitempty"`
+	PackageStatus string    `json:"packageStatus"`
+	SchemaURL     string    `json:"schemaURL"`
+	CreatedAt     time.Time `json:"createdAt"`
+}
+
+// PackageListResponse represents the API response structure for package lists
+type PackageListResponse struct {
+	Packages []PackageMetadata `json:"packages"`
+}
+
+// NewFileSystemProvider creates a new FileSystemProvider
+func NewFileSystemProvider(registryDir string) *FileSystemProvider {
+	return &FileSystemProvider{
+		registryDir: registryDir,
+	}
+}
```

Comment:
Nit: This should return a `PackageMetadataProvider` and `FileSystemProvider` should be private. Likewise with `NewAPIProvider`.